OCI Registries: Use separate auth tokens for each repository, rather than just one auth token per registry

### Community note

> [!TIP]
> 👋 Hi there, OpenTofu community! The OpenTofu team prioritizes issues based on upvotes. Please make sure to upvote this issue and describe how it affects you in detail in the comments to show your support.


### OpenTofu Version

```shell
OpenTofu v1.10.6
```

### OpenTofu Configuration Files

```hcl
module "eks" {
  source = "oci://harbor.example.com/project/tfmod-aws-eks?tag=v1.0"
  cluster = "example-eks"
}
```


### Debug Output

OpenTofu debug logs -

```
2025-09-29T17:00:25.576Z [TRACE] ModuleInstaller: eks address "oci://harbor.example.com/project/tfmod-aws-eks?tag=v1.0" will be handled by go-getter
Downloading oci://harbor.example.com/project/tfmod-aws-eks?tag=v1.0 for eks...
2025-09-29T17:00:25.576Z [TRACE] getmodules: fetching "oci://harbor.example.com/project/tfmod-aws-eks?tag=v1.0" to ".terraform/modules/eks"
2025-09-29T17:00:25.577Z [TRACE] OCI credentials discovery reading /home/vscode/.config/containers/auth.json
2025-09-29T17:00:25.577Z [TRACE] OCI credentials discovery reading /home/vscode/.docker/config.json
2025-09-29T17:00:25.578Z [TRACE] OCI credentials discovery reading /home/vscode/.dockercfg
2025-09-29T17:00:25.578Z [DEBUG] OCI registry client will consider credentials from /home/vscode/.docker/config.json
2025-09-29T17:00:25.579Z [TRACE] HTTP client GET request to https://harbor.example.com/v2/
2025-09-29T17:00:25.866Z [DEBUG] Executing docker-style credentials helper "dev-containers-c9219a27-feac-4e4e-8422-538330db1e63" for https://harbor.example.com
2025-09-29T17:00:26.327Z [ERROR] Docker-style credential helper "dev-containers-c9219a27-feac-4e4e-8422-538330db1e63" failed for https://harbor.example.com: exit status 1
2025-09-29T17:00:26.328Z [TRACE] HTTP client GET request to https://harbor.example.com/service/token?service=harbor-registry
2025-09-29T17:00:26.358Z [TRACE] modsdir: writing modules manifest to .terraform/modules/modules.json
╷
│ Error: Failed to download module
│ 
│   on main.tf line 1:
│    1: module "eks" {
│ 
│ Could not download module "eks" (main.tf:1) source code from "oci://harbor.example.com/project/tfmod-aws-eks?tag=v1.0": error downloading
│ 'oci://harbor.example.com/project/tfmod-aws-eks?tag=v1.0': configuring client for harbor.example.com/project/tfmod-aws-eks: failed to contact OCI
│ registry at "harbor.example.com": GET "https://harbor.example.com/v2/": GET
│ "https://harbor.example.com/service/token?service=harbor-registry": response status code 401: Unauthorized
╵
```

From Harbor logs
```
"GET /service/token?service=harbor-registry HTTP/1.1" 401 0 "-" "OpenTofu/1.10.6" 0.005 0.005 .
```

### Expected Behavior

Module is downloaded correctly as long as - 
1. harbor.example.com is accessible from the machine running opentofu, and
2. `project/tfmod-aws-eks` is a public repository which allows anonymous pulls 

### Actual Behavior

```
Could not download module "eks" (main.tf:1) source code from "oci://harbor.example.com/project/tfmod-aws-eks?tag=v1.0": error downloading 'oci://harbor.example.com/project/tfmod-aws-eks?tag=v1.0': configuring client for harbor.example.com/project/tfmod-aws-eks: failed to contact OCI registry at "harbor.example.com": GET "https://harbor.example.com/v2/": GET "https://harbor.example.com/service/token?service=harbor-registry": response status code 401: Unauthorized
```

### Steps to Reproduce

`tofu init`

### Additional Context

We have a private self-hosted Harbor registry running in our environment. This registry instance hosts multiple _public_ projects (public simply means projects that allow anonymous pulls, not that it is internet facing). 

All repositories within these internal _public_ projects are accessible anonymously. This setup works fine when pulling images. However, we get this 401 unauthorized error when trying to use a public project/repo to pull OpenTofu modules.

The following call to fetch the token succeeds, note the additional query parameters specifying `scope=repository/project/module`. This is absent in the OpenTofu call, and likely the cause of 401 errors.

```
curl "https://harbor.example.com/service/token?scope=repository%3Aproject%3Atfmod-aws-eks&service=harbor-registry" 
#
"GET /service/token?scope=repository% 3Aproject%3Atfmod-aws-eks&service=harbor-registry HTTP/1.1" 200 805 "-" "curl/8.5.0" 0.007 0.007 .
```


### References

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,7 @@ ENHANCEMENTS:
 
 - The `cidrsubnets` function now supports prefix extensions greater than 32 bits when the base CIDR block uses an IPv6 address. ([#4042](https://github.com/opentofu/opentofu/pull/4042))
 - The `local-exec` provisioner now automatically sets the `TRACEPARENT` environment variable in child processes when OpenTelemetry tracing is active, following the W3C Trace Context specification. ([#4014](https://github.com/opentofu/opentofu/issues/4014))
+- When installing provider and module packages from OCI Distribution registries, OpenTofu now tracks separate transient credentials for each repository to support registry implementations that issue repository-scoped tokens.  ([#3316](https://github.com/opentofu/opentofu/issues/3316))
 
 BUG FIXES:
 

--- a/cmd/tofu/oci_distribution.go
+++ b/cmd/tofu/oci_distribution.go
@@ -105,11 +105,6 @@ func getOCIRepositoryStore(ctx context.Context, registryDomain, repositoryName s
 		return nil, err // This is only for registryDomain validation errors, and we should've caught those much earlier than here
 	}
 	reg.Client = client
-	err = reg.Ping(ctx) // tests whether the given domain refers to a valid OCI repository and will accept the credentials
-	if err != nil {
-		tracing.SetSpanError(span, err)
-		return nil, fmt.Errorf("failed to contact OCI registry at %q: %w", registryDomain, err)
-	}
 	repo, err := reg.Repository(ctx, repositoryName)
 	if err != nil {
 		tracing.SetSpanError(span, err)

--- a/cmd/tofu/oci_distribution.go
+++ b/cmd/tofu/oci_distribution.go
@@ -105,6 +105,14 @@ func getOCIRepositoryStore(ctx context.Context, registryDomain, repositoryName s
 		return nil, err // This is only for registryDomain validation errors, and we should've caught those much earlier than here
 	}
 	reg.Client = client
+	// NOTE: We intentionally do not call reg.Ping(ctx) here.
+	// The Ping method targets the GET /v2/ endpoint, which some registries
+	// respond to with a 401 challenge, even for repositories with an
+	// anonymous pull configuration, which is interpreted as an error
+	//
+	// Since ORAS handles authentication transparently on the first real
+	// API call, the Ping is both redundant and problematic for this type of
+	// registry.
 	repo, err := reg.Repository(ctx, repositoryName)
 	if err != nil {
 		tracing.SetSpanError(span, err)

--- a/cmd/tofu/oci_distribution_test.go
+++ b/cmd/tofu/oci_distribution_test.go
@@ -59,8 +59,6 @@ func TestOCICredentialsLookupEnv_DockerCredHelper(t *testing.T) {
 // Test that getOCIRepositoryORASClient uses the correct credentials for each repository
 // when a registry has multiple repositories with different credentials.
 func TestGetOCIRepositoryORASClient_PerRepositoryCredentials(t *testing.T) {
-	ctx := t.Context()
-
 	// Configure two repositories with different credentials for same registry
 	credsPolicy := ociauthconfig.NewCredentialsConfigs([]ociauthconfig.CredentialsConfig{
 		&testPerRepoCredentialsConfig{
@@ -73,24 +71,31 @@ func TestGetOCIRepositoryORASClient_PerRepositoryCredentials(t *testing.T) {
 	})
 
 	for _, id := range [...]string{"a", "b"} {
-		client, err := getOCIRepositoryORASClient(ctx, "registry.example.com", "repo-"+id, credsPolicy)
-		if err != nil {
-			t.Fatalf("creating client for repo-%s: %s", id, err)
-		}
+		t.Run(id, func(t *testing.T) {
+			ctx := t.Context()
+			client, err := getOCIRepositoryORASClient(ctx, "registry.example.com", "repo-"+id, credsPolicy)
 
-		credentials, err := client.Credential(ctx, "registry.example.com")
-		if err != nil {
-			t.Fatalf("resolving credential for repo-%s: %s", id, err)
-		}
+			if err != nil {
+				t.Fatalf("failed creating credential: %s", err)
+			}
 
-		if credentials.Username != "user-"+id {
-			t.Errorf("repo-%s: wrong username: got %s, want %s", id, credentials.Username, "user-"+id)
-		}
-		if credentials.Password != "password-"+id {
-			t.Errorf("repo-%s: wrong password: got %s, want %s", id, credentials.Password, "password-"+id)
-		}
+			credentials, err := client.Credential(ctx, "registry.example.com")
+			if err != nil {
+				t.Fatalf("failed resolving credential: %s", err)
+			}
+
+			want, got := "user-"+id, credentials.Username
+			if want != got {
+				t.Fatalf("username mismatch: %s != %s", got, want)
+			}
+
+			want, got = "password-"+id, credentials.Password
+			if want != got {
+				t.Fatalf("password mismatch: %s != %s", got, want)
+			}
+
+		})
 	}
-
 }
 
 // testPerRepoCredentialsConfig is a test-only implementation of

--- a/cmd/tofu/oci_distribution_test.go
+++ b/cmd/tofu/oci_distribution_test.go
@@ -13,6 +13,8 @@ import (
 	"testing"
 
 	"github.com/opentofu/opentofu/internal/command/cliconfig/ociauthconfig"
+	"github.com/opentofu/opentofu/internal/command/e2etest/fakeocireg"
+	orasRemote "oras.land/oras-go/v2/registry/remote"
 )
 
 func TestOCICredentialsLookupEnv_DockerCredHelper(t *testing.T) {
@@ -95,6 +97,52 @@ func TestGetOCIRepositoryORASClient_PerRepositoryCredentials(t *testing.T) {
 			}
 
 		})
+	}
+}
+
+// TestGetOciRepository verifies that getOCIRepositoryStore can successfully
+// connect to an OCI registry that returns a 401 Unauthorized response for
+// the `GET /v2/` endpoint, which is a behavior seen with some registries.
+//
+// This test ensures that OpenTofu does not use the problematic `oras.Ping()`
+// function, which would fail against such registries. Instead, authentication
+// is handled transparently by ORAS on the first real API call.
+func TestGetOciRepository(t *testing.T) {
+	ctx := t.Context()
+
+	// No authentication, empty credentials
+	credsPolicy := ociauthconfig.NewCredentialsConfigs(nil)
+
+	registryServer, err := fakeocireg.NewServer(t.Context(), map[string]string{	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer registryServer.Close()
+
+	registryAddr := registryServer.Listener.Addr().String()
+	repositoryName := "repo-a"
+
+	store, err := getOCIRepositoryStore(ctx, registryAddr, repositoryName, credsPolicy)
+
+	if err != nil {
+		t.Fatalf("failed creating oci repository store: %s", err)
+	}
+
+	if store == nil {
+		t.Fatal("failed creating oci repository store, expected non-nil store")
+	}
+
+	repo, ok := store.(*orasRemote.Repository)
+	if !ok {
+		t.Fatal("failed creating oci repository store, expected orasRemote.Repository")
+	}
+
+	if repo.Reference.Registry != registryAddr {
+		t.Errorf("wrong registry; got %q, want %q", repo.Reference.Registry, registryAddr)
+	}
+
+	if repo.Reference.Repository != repositoryName {
+		t.Errorf("wrong repository name; got %q, want %q", repo.Reference.Repository, repositoryName)
 	}
 }
 

--- a/cmd/tofu/oci_distribution_test.go
+++ b/cmd/tofu/oci_distribution_test.go
@@ -100,14 +100,14 @@ func TestGetOCIRepositoryORASClient_PerRepositoryCredentials(t *testing.T) {
 	}
 }
 
-// TestGetOciRepository verifies that getOCIRepositoryStore can successfully
+// TestGetOciRepositoryNoPing verifies that getOCIRepositoryStore can successfully
 // connect to an OCI registry that returns a 401 Unauthorized response for
 // the `GET /v2/` endpoint, which is a behavior seen with some registries.
 //
 // This test ensures that OpenTofu does not use the problematic `oras.Ping()`
 // function, which would fail against such registries. Instead, authentication
 // is handled transparently by ORAS on the first real API call.
-func TestGetOciRepository(t *testing.T) {
+func TestGetOciRepositoryNoPing(t *testing.T) {
 	ctx := t.Context()
 
 	// No authentication, empty credentials

--- a/cmd/tofu/oci_distribution_test.go
+++ b/cmd/tofu/oci_distribution_test.go
@@ -6,9 +6,13 @@
 package main
 
 import (
+	"context"
+	"iter"
 	"os"
 	"strings"
 	"testing"
+
+	"github.com/opentofu/opentofu/internal/command/cliconfig/ociauthconfig"
 )
 
 func TestOCICredentialsLookupEnv_DockerCredHelper(t *testing.T) {
@@ -50,4 +54,66 @@ func TestOCICredentialsLookupEnv_DockerCredHelper(t *testing.T) {
 	if gotErr := err.Error(); !strings.Contains(gotErr, wantErr) {
 		t.Errorf("wrong error\ngot: %s\nwant substring: %s", gotErr, wantErr)
 	}
+}
+
+// Test that getOCIRepositoryORASClient uses the correct credentials for each repository
+// when a registry has multiple repositories with different credentials.
+func TestGetOCIRepositoryORASClient_PerRepositoryCredentials(t *testing.T) {
+	ctx := t.Context()
+
+	// Configure two repositories with different credentials for same repository
+	credsPolicy := ociauthconfig.NewCredentialsConfigs([]ociauthconfig.CredentialsConfig{
+		&testPerRepoCredentialsConfig{
+			domain: "registry.example.com",
+			repos: map[string]ociauthconfig.Credentials{
+				"repo-a": ociauthconfig.NewBasicAuthCredentials("user-a", "password-a"),
+				"repo-b": ociauthconfig.NewBasicAuthCredentials("user-b", "password-b"),
+			},
+		},
+	})
+
+	for _, id := range [...]string{"a", "b"} {
+		client, err := getOCIRepositoryORASClient(ctx, "registry.example.com", "repo-"+id, credsPolicy)
+		if err != nil {
+			t.Fatalf("creating client for repo-%s: %s", id, err)
+		}
+
+		credentials, err := client.Credential(ctx, "registry.example.com")
+		if err != nil {
+			t.Fatalf("resolving credential for repo-%s: %s", id, err)
+		}
+
+		if credentials.Username != "user-"+id {
+			t.Errorf("repo-a: wrong username: got %q, want %q", credentials.Username, "user-"+id)
+		}
+		if credentials.Password != "password-"+id {
+			t.Errorf("repo-a: wrong password: got %q, want %q", credentials.Password, "password-"+id)
+		}
+	}
+
+}
+
+// testPerRepoCredentialsConfig is a test-only implementation of
+// [ociauthconfig.CredentialsConfig] that maps repository paths within a fixed
+// domain to static credentials.
+type testPerRepoCredentialsConfig struct {
+	domain string
+	repos  map[string]ociauthconfig.Credentials
+}
+
+func (c *testPerRepoCredentialsConfig) CredentialsSourcesForRepository(_ context.Context, registryDomain, repositoryPath string) iter.Seq2[ociauthconfig.CredentialsSource, error] {
+	return func(yield func(ociauthconfig.CredentialsSource, error) bool) {
+		if registryDomain != c.domain {
+			return
+		}
+		creds, ok := c.repos[repositoryPath]
+		if !ok {
+			return
+		}
+		yield(ociauthconfig.NewStaticCredentialsSource(creds, ociauthconfig.RepositoryCredentialsSpecificity(1)), nil)
+	}
+}
+
+func (c *testPerRepoCredentialsConfig) CredentialsConfigLocationForUI() string {
+	return "test fixture"
 }

--- a/cmd/tofu/oci_distribution_test.go
+++ b/cmd/tofu/oci_distribution_test.go
@@ -61,7 +61,7 @@ func TestOCICredentialsLookupEnv_DockerCredHelper(t *testing.T) {
 func TestGetOCIRepositoryORASClient_PerRepositoryCredentials(t *testing.T) {
 	ctx := t.Context()
 
-	// Configure two repositories with different credentials for same repository
+	// Configure two repositories with different credentials for same registry
 	credsPolicy := ociauthconfig.NewCredentialsConfigs([]ociauthconfig.CredentialsConfig{
 		&testPerRepoCredentialsConfig{
 			domain: "registry.example.com",
@@ -84,10 +84,10 @@ func TestGetOCIRepositoryORASClient_PerRepositoryCredentials(t *testing.T) {
 		}
 
 		if credentials.Username != "user-"+id {
-			t.Errorf("repo-a: wrong username: got %q, want %q", credentials.Username, "user-"+id)
+			t.Errorf("repo-%s: wrong username: got %s, want %s", id, credentials.Username, "user-"+id)
 		}
 		if credentials.Password != "password-"+id {
-			t.Errorf("repo-a: wrong password: got %q, want %q", credentials.Password, "password-"+id)
+			t.Errorf("repo-%s: wrong password: got %s, want %s", id, credentials.Password, "password-"+id)
 		}
 	}
 

--- a/internal/command/e2etest/fakeocireg/handler.go
+++ b/internal/command/e2etest/fakeocireg/handler.go
@@ -44,8 +44,15 @@ func (r registryHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) 
 	}
 	if repoName == "" {
 		// This is just a "does this server even support the protocol?" discovery request.
+		// The OCI specification includes a `GET /v2/` endpoint for API version checks. Some
+		// registries incorrectly require authentication for this endpoint, even for anonymous
+		// pulls.
+		//
+		// To prevent authentication problems, OpenTofu must not use this endpoint.
+		// We intentionally fail this request to ensure that tests break if this
+		// endpoint is ever used, for example via `oras.Ping`.
 		log.Printf("[INFO] fakeocireg: protocol discovery succeeds")
-		resp.WriteHeader(http.StatusOK)
+		resp.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 	store, ok := r.stores[repoName]

